keps/sig-api-machinery/xxx-resilient-watchcache-initialization/README.md

```diff
@@ -0,0 +1,764 @@
+of a given type, initialiation (being effectively a list from etcd) can take
+tens of seconds or even above a minute and in the meantime this watch is consuming
+seats of API Priority&Fairness. In many cases at most tens of such watches (which
+is a relatively small number in clusters with thousands of nodes) can effectively
+completely starve a given PriorityLevel and block any load on it (both write and
+read requests).
+
+For get/list requests that explicitly opt-in for being served from watchcache
+(namely those marked as `NotOlderThan` or `Any` in [the documentation]) - which
+amongst others is what Reflector/Informers framework is using in most of cases,
+if watchcache is not initialized there are two cases:
+- if ResourceVersion=0, these are blindly delegated to etcd
+- if ResourceVersion>0, they hang waiting for initialization similarly as watch
+requests
+
+The first case is particularly problematic for lists that hugely benefit from
+watchcache indexing (e.g. listing pods assigned to a given node). When served
+from watchcache, these requests are really efficient, but serving them from etcd
+requires reading a large amount of data, deserializing it, filtering out objects
+not matching selector and sending only those to the client. Passing too many of
+such requests down to etcd, can easily bring down the whole control plane.
+Additionally, in this case, when served from watchcache a LIMIT parameter is
+ignored and the whole result is returned in one go. If served from etcd, the
+LIMIT parameter is honored, so for large collections we end-up paging the
+response (which for large collections and limit=500 of reflector/informer
+frameworks can result in hundreds of consecutive requests for next pages).
+
+The second case is suffering from the same problem as watches, as those requests
+may effectively starve the whole PriorityLevel.
+
+In order to mitigate the above problem, we propose a couple different changes.
+
+[the documentation]: https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list
+
+#### Reduce the number of requests during initialization
+
+As a first step, we will try to reduce the number of requests that are even reaching
+kube-apiserver with unitialized watchcache. To achieve it, we will introduce a new
+PostStartHook that will wait for watchcache of all *builtin* resources to be initialized.
+
+Why this should help with reducing the number of requests? The whole idea behind
+kube-apiserver readiness is that Kubernetes installations should be configured so that
+requests are not sent to it until kube-apiserver becomes ready. The above will allow us
+to delay this moment until watchcache is initialized.
+
+Why PostStartHook instead of /readyz check? Watchcache is per-type layer and we want
+to avoid marking whole kube-apiserver as not-ready if watchcache for one of resources
+requires reinitialization. We will handle those cases differently.
```

Comment:
Just making sure I understand.  Would the first initialization of each builtin type be /readyz gated but any re-initialization be post- /readyz (but still result in 429 responses)?

Reply:
The way poststarthook works is just it effectively:
1) is starteded once:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/hooks.go#L194-L206
2) it's reported as healthy as soon as it succeeds for the first time:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/hooks.go#L238-L245
[and it never changes back to false later]

So what will happen is effectively:
- first initialization of each builtin type will prevent readyz from succeeding [initialization of kube-apiserver]
- even if watchcache needs to reinitialize later, it will not bring readyz/ back to failing, but as you wrote, requests [which exactly described below] will be returning 429

tried to clarify in the text